File: common/model/holding.ts
import { Account } from "@common";
import { DBBase } from "./base";

/** This holding class keeps track of an investment holding value */
export class Holding extends DBBase {
  currency: string;
  costBasis: number;
  /** A description of what this holding is */
  description: string;
  /** The current market value */
  marketValue: number;
  /** The current purchase price */
  purchasePrice: number;
  /** Total number of shares, including fractional */
  shares: number;
  /** The symbol for this holding */
  symbol: string;

  /** The account this holding is associated to */
  account: Account;

  constructor(
    currency: string,
    costBasis: number,
    description: string,
    marketValue: number,
    purchasePrice: number,
    shares: number,
    symbol: string,
    account: Account,
  ) {
    super();
    this.currency = currency;
    this.costBasis = costBasis;
    this.description = description;
    this.marketValue = marketValue;
    this.purchasePrice = purchasePrice;
    this.shares = shares;
    this.symbol = symbol;
    this.account = account;
  }
}

File: frontend/lib/holding/models/holding.dart
import 'package:sprout/account/models/account.dart';

/// This class provides information for a current stock that is associated to an account.
class Holding {
  /// The account this holding is associated to
  final Account account;

  final String currency;

  final double costBasis;

  /// A description of what this holding is
  final String description;

  /// The current market value
  final double marketValue;

  /// The current purchase price
  final double purchasePrice;

  /// Total number of shares, including fractional
  final double shares;

  /// The symbol for this holding
  final String symbol;

  Holding({
    required this.currency,
    required this.costBasis,
    required this.description,
    required this.marketValue,
    required this.purchasePrice,
    required this.shares,
    required this.symbol,
    required this.account,
  });

  factory Holding.fromJson(Map<String, dynamic> json) {
    return Holding(
      currency: json['currency'],
      costBasis: json['costBasis']?.toDouble(),
      description: json['description'],
      marketValue: json['marketValue']?.toDouble(),
      purchasePrice: json['purchasePrice']?.toDouble(),
      shares: json['shares']?.toDouble(),
      symbol: json['symbol'],
      account: Account.fromJson(json['account']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'currency': currency,
      'costBasis': costBasis,
      'description': description,
      'marketValue': marketValue,
      'purchasePrice': purchasePrice,
      'shares': shares,
      'symbol': symbol,
      'account': account.toJson(),
    };
  }
}
